File: src/test/java/br/com/blz/testjava/service/ProductServiceTest.java
package br.com.blz.testjava.service;

import br.com.blz.testjava.dto.InventoryDTO;
import br.com.blz.testjava.dto.ProductDTO;
import br.com.blz.testjava.dto.WarehouseDTO;
import br.com.blz.testjava.model.InventoryEntity;
import br.com.blz.testjava.model.ProductEntity;
import br.com.blz.testjava.model.WarehouseEntity;
import br.com.blz.testjava.repository.ProductRepositoryImpl;
import org.junit.Assert;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.ArrayList;
import java.util.List;

@ExtendWith(MockitoExtension.class)
public class ProductServiceTest {

    @InjectMocks
    private ProductService service;

    @Mock
    private ProductRepositoryImpl repository;

    @Test
    public void create_new_product(){
//        Integer sku = 1564;
//
//        //Mockito.when(repository.findProductBySku(sku)).thenReturn(buildProduct());
//        //Mockito.doCallRealMethod().when(repository.findProductBySku(buildProduct().getSku()));
//        service.createProduct(buildProductDTO());
//
//        //Mockito.verify(repository, Mockito.only()).findProductBySku(sku);
//        Mockito.verify(repository, Mockito.only()).createProduct(buildProduct());
//
//        Assertions.assertEquals(sku, buildProduct().getSku());
//        //Assertions.assertNotNull(repository.findProductBySku(sku));



    }

    private ProductEntity buildProduct(){
        WarehouseEntity wareHouseEccomerce = new WarehouseEntity();
        wareHouseEccomerce.setLocality("SP");
        wareHouseEccomerce.setQuantity(2);
        wareHouseEccomerce.setType("ECOMMERCE");

        WarehouseEntity wareHousePhysicalStore = new WarehouseEntity();
        wareHousePhysicalStore.setLocality("SP");
        wareHousePhysicalStore.setQuantity(1);
        wareHousePhysicalStore.setType("PHYSICAL_STORE");

        List<WarehouseEntity> warehouses = new ArrayList<>();
        warehouses.add(wareHouseEccomerce);
        warehouses.add(wareHousePhysicalStore);

        InventoryEntity inventory = new InventoryEntity();
        inventory.setWarehouses(warehouses);

        ProductEntity product = new ProductEntity();
        product.setSku(1564);
        product.setName("L'oreal");
        product.setInventory(inventory);

        return product;
    }

    private ProductDTO buildProductDTO(){
        WarehouseDTO wareHouseEccomerce = new WarehouseDTO();
        wareHouseEccomerce.setLocality("SP");
        wareHouseEccomerce.setQuantity(2);
        wareHouseEccomerce.setType("ECOMMERCE");

        WarehouseDTO wareHousePhysicalStore = new WarehouseDTO();
        wareHousePhysicalStore.setLocality("SP");
        wareHousePhysicalStore.setQuantity(1);
        wareHousePhysicalStore.setType("PHYSICAL_STORE");

        List<WarehouseDTO> warehouses = new ArrayList<>();
        warehouses.add(wareHouseEccomerce);
        warehouses.add(wareHousePhysicalStore);

        InventoryDTO inventory = new InventoryDTO();
        inventory.setWarehouses(warehouses);

        ProductDTO product = new ProductDTO();
        product.setSku(1564);
        product.setName("L'oreal");
        product.setInventory(inventory);

        return product;
    }
}

File: src/test/java/br/com/blz/testjava/service/ProductServiceTest.kt
package br.com.blz.testjava.service

import br.com.blz.testjava.ProductTestData
import br.com.blz.testjava.exception.ProductAlreadyExistsException
import br.com.blz.testjava.exception.ProductNotFoundException
import br.com.blz.testjava.model.Product
import br.com.blz.testjava.repository.ProductRepository
import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class ProductServiceTest {

  @MockK
  private lateinit var productRepository: ProductRepository

  @InjectMockKs
  private lateinit var productService: ProductService

  @BeforeEach
  fun setUp() {
    unmockkAll()
    MockKAnnotations.init(this)
  }


  @Test
  fun should_not_get_product_not_found() {
    val sku: Long = 43264

    ProductTestData.productNotMarketable()

    every { productRepository.get(sku) } returns null

    assertThrows<ProductNotFoundException> { productService.getProductBySku(sku) }
  }

  @Test
  fun should_mark_product_as_not_marketable_and_0_inventory_quantity() {
    val sku: Long = 43264

    val productNotMarketable = ProductTestData.productNotMarketable()

    every { productRepository.get(sku) } returns productNotMarketable

    val response = productService.getProductBySku(sku)
    assertFalse(response.isMarketable())
    assertEquals(0, response.inventory.quantity)
  }

  @Test
  fun should_mark_product_as_marketable_and_with_15_inventories() {
    val sku: Long = 43264

    val product = ProductTestData.productMarketable()

    every { productRepository.get(sku) } returns product

    val response = productService.getProductBySku(sku)
    assertTrue(response.isMarketable())
    assertEquals(15, response.inventory.quantity)
  }

  @Test
  fun should_create_new_product_sucessfully() {
    val sku: Long = 43264

    val product: Product = ProductTestData.productMarketable()!!

    every { productRepository.get(sku) } returns null
    every { productRepository.create(any()) } returns product

    productService.createProduct(product)

    verify(exactly = 1) { productRepository.create(product) }
  }


  @Test
  fun should_not_create_new_product_because_already_exists() {
    val sku: Long = 43264

    val product: Product = ProductTestData.productMarketable()!!

    every { productRepository.get(sku) } returns mockk()

    assertThrows<ProductAlreadyExistsException> { productService.createProduct(product) }

    verify(exactly = 0) { productRepository.create(product) }
  }

  @Test
  fun should_update_product() {

    val sku: Long = 43264

    val product: Product = ProductTestData.productMarketable()!!

    every { productRepository.update(sku, product) } returns mockk()

     productService.updateProduct(sku, product)

    verify(exactly = 1) { productRepository.update(sku, product) }
  }

  @Test
  fun should_delete_product() {

    val sku: Long = 43264

    ProductTestData.productMarketable()!!

    every { productRepository.delete(sku) } returns mockk()

    productService.deleteProduct(sku)

    verify(exactly = 1) { productRepository.delete(sku) }
  }

}
